src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,171 @@
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            head = new Node<T>(null, value, tail);
+            tail = new Node<T>(head, value, null);
+        }
+        if (size == 1) {
+            tail = new Node<>(head, value, null);
+            head.next = tail;
+        }
+        if (size > 1) {
+            Node<T> main = new Node<>(tail, value, null);
+            tail.next = main;
+            tail = main;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (size == index) {
+            add(value);
+            return;
+        }
+        checkIndexInBounds(index);
+        addBeforeIndex(value, index);
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndexInBounds(index);
+        Node<T> node = getNode(index);
+        if (node != null) {
+            return node.main;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (!isEmpty() || index < 0) {
+            checkIndexInBounds(index);
+        }
+        Node<T> node = getNode(index);
+        if (node != null) {
+            T result = node.main;
+            node.main = value;
+            return result;
+        }
+        return value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (!isEmpty() || index < 0) {
+            checkIndexInBounds(index);
+        }
+        Node<T> result = getNode(index);
+        assert result != null;
+        unlink(result);
+        size--;
+        return result.main;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        int curentIndex = 0;
+        while (node != null) {
+            if (isEquals(object, node.main)) {
+                unlink(node);
+                size--;
+                return true;
+            }
+            node = node.next;
+            curentIndex++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
+        int currentIndex = 0;
+        Node<T> current = head;
+        while (current != null) {
+            if (index == currentIndex) {
+                return current;
+            }
+            current = current.next;
+            currentIndex++;
+        }
+        return null;
+    }
+
+    private void addBeforeIndex(T value, int index) {
+        Node<T> node = getNode(index);
+        Node<T> nodeValue = new Node<>(null, value, null);
+        assert node != null;
```

Comment:
if index is not valid checkIndexInBounds would throw an exception
and if it doesnt - there will always be a node by the given index